### PR TITLE
[libc++][test] Refactor increasing_allocator

### PR DESCRIPTION
**Pull request by winner245**
The `increasing_allocator<T>` class, originally introduced to test `shrink_to_fit` for `std::vector`, `vector<bool>`, and `std::basic_string` (#95161), has duplicated definitions across several test files. Given the potential utility of this class for capacity-related tests in various sequence containers, this PR proposes refactoring the definition of `increasing_allocator<T>` into a single, reusable location.

### CODE REVIEW

File: libcxx/test/std/containers/sequences/vector.bool/shrink_to_fit.pass.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ struct increasing_allocator {
     min_elements += 1000;
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/shrink_to_fit.pass.cpp:55]: Can we factor out the code into a single location?

**Comment by philnik777** [comment]
I feel like the reasoning is backwards here. AFAICT the only problem is that we're mismatching the `n` we pass to `allocator::deallocate` with the `n` we're passing to `allocator::allocate`. Nothing requires that there is space for exactly n elements in an allocation - there has to be space for at least that. In fact, most allocators return more than the requested amount in at least some cases, which is the whole reason `allocate_at_least` was introduced.

**Comment by ldionne** [comment]
@winner245 Actually, how did you come across this issue? Did our test suite fail with another Standard library somehow?

**Comment by winner245** [comment]
Because I just realized that the same issue #95161 applies to `__split_buffer::shrink_to_fit`, which can increase the capacity of the underlying buffer when the allocator's `allocate_at_least` allocates more memory than requested. 

I was trying to solve a different way to solve the `shrink_to_fit` issue because I think the root cause to all these `shink_to_fit` issues is that we used `allocate_at_least` instead of `allocate`. If we just revert back to `allocate`, I believe all the `shinrk_to_fit` issue would automatically be gone. This is because the standard ([[allocator.requirements]/36](https://eel.is/c++draft/allocator.requirements#general-36)) mandates that `a.allocate(n)` must allocate memory for exactly n objects of type T.

**Comment by winner245** [comment]
I personally do not like the current solution, because `allocate_at_least` may end up allocating more memory than requested. Then we would have to immediately discard the newly allocated memory. If we revert back to `allocate`, we wouldn't have the problem.

**Comment by philnik777** [comment]
> I personally do not like the current solution, because `allocate_at_least` may end up allocating more memory than requested. Then we would have to immediately discard the newly allocated memory. If we revert back to `allocate`, we wouldn't have the problem.

The intention of `allocate_at_least` is that the allocator tells us how much it overallocated due to its allocation algorithm, not that it actually allocates more. If you implement the interface properly `allocate` will only hide the overallocation, not prevent it.

**Comment by ldionne** [comment]
> The intention of `allocate_at_least` is that the allocator tells us how much it overallocated due to its allocation algorithm, not that it actually allocates more. If you implement the interface properly `allocate` will only hide the overallocation, not prevent it.

I agree, and I don't see how this makes the allocator more conforming than it used to be. However, I do think the code makes more sense after the patch than it did before the patch, because implementing `allocate()` using `std::allocator::allocate` makes more sense than doing so with `allocate_at_least`.


> Actually, the tests failed with MSVC STL (or MSVC, the compiler). MSVC detects size mismatch of deallocation in constant evaluation. I guess we can't reliably perform over allocation in constant evaluation.

```cpp
constexpr std::allocation_result<T*> allocate_at_least(std::size_t n);
constexpr T* allocate(std::size_t n) {
  auto res = allocate_at_least(n);
  return res.ptr;
}
constexpr void deallocate(T* p, std::size_t n) noexcept {
  // n may be wrong: it must match res.count which may not be the case if we used allocate_at_least above
  std::allocator<T>{}.deallocate(p, n);
}
```

Is that what you're saying? I would understand why the current code is a problem in that case, however http://eel.is/c++draft/allocator.members#10.1 says:

> If p is memory that was obtained by a call to allocate_at_least, let ret be the value returned and req be the value passed as the first argument to that call[.](http://eel.is/c++draft/allocator.members#10.1.sentence-1) p is equal to ret.ptr and n is a value such that req  ≤ n  ≤ ret[.](http://eel.is/c++draft/allocator.members#10.1.sentence-2)count.

In other words, I think it's supposed to be just fine to call `std::allocator::deallocate` with a `n` that is less than the count returned by a previous `allocate_at_least`.

**Comment by philnik777** [comment]
> > The intention of `allocate_at_least` is that the allocator tells us how much it overallocated due to its allocation algorithm, not that it actually allocates more. If you implement the interface properly `allocate` will only hide the overallocation, not prevent it.
> 
> I agree, and I don't see how this makes the allocator more conforming than it used to be. However, I do think the code makes more sense after the patch than it did before the patch, because implementing `allocate()` using `std::allocator::allocate` makes more sense than doing so with `allocate_at_least`.

I'm not disagreeing here. The code is an improvement, but I don't think the commit message makes sense.

> > Actually, the tests failed with MSVC STL (or MSVC, the compiler). MSVC detects size mismatch of deallocation in constant evaluation. I guess we can't reliably perform over allocation in constant evaluation.
> 
> ```c++
> constexpr std::allocation_result<T*> allocate_at_least(std::size_t n);
> constexpr T* allocate(std::size_t n) {
>   auto res = allocate_at_least(n);
>   return res.ptr;
> }
> constexpr void deallocate(T* p, std::size_t n) noexcept {
>   // n may be wrong: it must match res.count which may not be the case if we used allocate_at_least above
>   std::allocator<T>{}.deallocate(p, n);
> }
> ```
> 
> Is that what you're saying? I would understand why the current code is a problem in that case, however http://eel.is/c++draft/allocator.members#10.1 says:
> 
> > If p is memory that was obtained by a call to allocate_at_least, let ret be the value returned and req be the value passed as the first argument to that call[.](http://eel.is/c++draft/allocator.members#10.1.sentence-1) p is equal to ret.ptr and n is a value such that req  ≤ n  ≤ ret[.](http://eel.is/c++draft/allocator.members#10.1.sentence-2)count.
> 
> In other words, I think it's supposed to be just fine to call `std::allocator::deallocate` with a `n` that is less than the count returned by a previous `allocate_at_least`.

The calling code is fine. The mismatch between the `std::allocator::allocate()` and `std::allocator::deallocate()` is the problem. Note how `increasing_allocator::allocate_at_least` calls `std::allocator::allocate(n + 1000)` in some cases, but `increasing_allocator::deallocate()` always just calls `std::allocator::deallocate(n)`.

Actually, I just noticed that the current code is still wrong, since there is still potentially the same mismatch when calling `increasing_allocator::allocate_at_least` and `increasing_allocator::deallocate`.

**Comment by winner245** [comment]
Thank you all for the very informative discussions here. I agree that the current solution for `shrink_to_fit` is the best compromise we have at this point. In response to @philnik777's comment that "allocate will only hide the overallocation, not prevent it," I think this PR now serves as a refactoring:

- Currently, the `increasing_allocator<T>` class appears in three different files. We should factor it out into a single location.
- Semantically, it makes more sense to let `allocate()` call `std::allocator::allocate`, not `allocate_at_least` (and  let `allocate_at_least()` call `std::allocator::allocate_at_least`).

File: libcxx/test/support/increasing_allocator.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  increasing_allocator()   = default;
+
+  template <typename U>
+  constexpr increasing_allocator(const increasing_allocator<U>& other) noexcept : min_elements(other.min_elements) {}
+
+  constexpr std::allocation_result<T*> allocate_at_least(std::size_t n) {
+    if (n < min_elements)
+      n = min_elements;
+    min_elements += 1000;
+    return std::allocator<T>{}.allocate_at_least(n);
```
Comment (frederick-vs-ja) [inline libcxx/test/support/increasing_allocator.h:31]: `std::allocator::allocate_at_least` (in all existing implementations) never performs over-allocation and just returns `{allocate(n), n}`. So the argument passed to the `allocator::deallocate` must be exactly the increased `n`. However, in order to meet the _Cpp17Allocator_ requirements, every value in `[old_n, new_n]` needs to be a valid argument for `increasing_allocator::deallocate`.

I guess if we want to emulate over-allocation with `std::allocator`, we should call `std::allocator<T>{}.allocate(DO_CEIL(n))` in **both** `allocate` and `allocate_at_least`, and call `std::allocator<T>{}.deallocate(p, DO_CEIL(n))` in `deallocate`, where `DO_CEIL` can be `std::bit_ceil` or something similar.

Comment (winner245) [reply libcxx/test/support/increasing_allocator.h:31]: This sounds like a very interesting approach to emulate over-allocation in `std::allocator::allocate_at_least`. Are you suggesting the following implementation?

```cpp
template <typename T>
struct bit_ceil_allocator {
  using value_type     = T; 
  bit_ceil_allocator() = default;

  template <typename U>
  constexpr bit_ceil_allocator(const bit_ceil_allocator<U>& other) noexcept {}

  constexpr std::allocation_result<T*> allocate_at_least(std::size_t n) { 
    n = std::bit_ceil(n);
    return std::allocator<T>{}.allocate_at_least(n);
  }

  constexpr T* allocate(std::size_t n) { 
    n = std::bit_ceil(n);
    return std::allocator<T>{}.allocate(n);
  }
  
  constexpr void deallocate(T* p, std::size_t n) noexcept { 
    std::allocator<T>{}.deallocate(p, std::bit_ceil(n)); 
  }
};
```

This implementation indeed performs over-allocation behind the scenes. It also appears to satisfy the _Cpp17Allocator_ requirements, which allows us to pass any value in `[n, std::bit_ceil(n)]` to `bit_ceil_allocator::deallocate`.

File: libcxx/test/support/increasing_allocator.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef INCREASING_ALLOCATOR_H
+#define INCREASING_ALLOCATOR_H
```
Comment (ldionne) [inline libcxx/test/support/increasing_allocator.h:10]:
```suggestion
#ifndef TEST_SUPPORT_INCREASING_ALLOCATOR_H
#define TEST_SUPPORT_INCREASING_ALLOCATOR_H
```

File: libcxx/test/support/increasing_allocator.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <cstddef>
+#include <memory>
+
+#include "test_macros.h"
+
+#if TEST_STD_VER >= 23
+template <typename T>
+struct increasing_allocator {
```
Comment (ldionne) [inline libcxx/test/support/increasing_allocator.h:19]: Could you please add a short comment explaining what this allocator does and in what ways it is special? Since this is promoted to a reusable utility, I think it's worth it.

**Review by ldionne — COMMENTED** [review]
It looks like we have agreement on the path forward. I left a few additional comments.

File: libcxx/test/support/increasing_allocator.h (hunk 4 above)
Comment (winner245) [reply libcxx/test/support/increasing_allocator.h:26]: Thank you for your suggestion! I have added a short comment explaining the purpose and special functionality of the `increasing_allocator`. Please let me know if you have further suggestions on it. 

I have also modified the PR description so that it reflects the real purpose of this PR.

File: libcxx/test/support/increasing_allocator.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  template <typename U>
+  constexpr increasing_allocator(const increasing_allocator<U>& other) noexcept : min_elements(other.min_elements) {}
+
+  constexpr std::allocation_result<T*> allocate_at_least(std::size_t n) {
+    if (n < min_elements)
+      n = min_elements;
+    min_elements += 1000;
+    return std::allocator<T>{}.allocate_at_least(n);
+  }
+  constexpr T* allocate(std::size_t n) { return std::allocator<T>{}.allocate(n); }
```
Comment (ldionne) [inline libcxx/test/support/increasing_allocator.h:40]: I think we still need to make sure we always allocate an increasing amount of memory in `allocate(...)`, otherwise this test allocator would not help catch the case where `shrink_to_fit` is implemented incorrectly but it uses `alloc.allocate` instead of `alloc.allocate_at_least`.

File: libcxx/test/support/increasing_allocator.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      n = min_elements;
+    min_elements += 1000;
+    return std::allocator<T>{}.allocate_at_least(n);
+  }
+#endif // TEST_STD_VER >= 23
+
+  TEST_CONSTEXPR_CXX20 T* allocate(std::size_t n) {
+#if TEST_STD_VER >= 23
+    return allocate_at_least(n).ptr;
+#else
```
[the remark this answers is not in the document]
Comment (winner245) [reply libcxx/test/support/increasing_allocator.h:45]: Thank you for your detailed feedback!

> Why don't we unconditionally call `std::allocator<T>().allocate(n)`?

I agree that it's better to unconditionally call `std::allocator<T>().allocate(n)` within `increasing_allocator::allocate(n)`. I also think that we should allow only the `allocate_at_least(n)` function of `increasing_allocator` to allocate more memory than the requested amount `n`, i.e., it actually allocates `max(n, min_elements)`. However, I think it makes more sense for `increasing_allocator::allocate(n)` to allocate effectively the exact amount `n`, otherwise we will see a mismatch between the reported and the actual allocation counts, as shown in this [example](https://godbolt.org/z/efGzqKb93). Therefore, `allocate_at_least(n)` should allocate at least `n`, while `allocate(n)` should allocate exactly `n`. 

> Also, we still run into the issue that we had before, which is that we're calling std::allocator<T>().allocate(n) but calling std::allocator<T>().deallocate(ptr, m) with a smaller value than we allocated. That violates [allocator.members](http://eel.is/c++draft/allocator.members#10.1).
> 
> We can either do what @frederick-vs-ja suggested [here](https://github.com/llvm/llvm-project/pull/115671/files#r1843551032) (which TBH I don't fully understand why that makes the allocator ever-increasing), or we could try something like this: [godbolt.org/z/P8PK6hvEj](https://godbolt.org/z/P8PK6hvEj)

I agree that the allocation/deallocation-size mismatch between `allocate(n)/allocate_at_least(n)` and `deallocate(p, m)` always exists because it completely depends on the usages of the functions. Users may provide mismatched `m/n` values, regardless of our implementation. I think @frederick-vs-ja's idea provides a partial solution or it alleviates this issue by quantizing the allocation/deallocation sizes into discrete power-of-2 values. This allows for mismatches between `n` and `m`, as long as they are quantized to the same power of 2. However, completely irrelevant `m, n` values yielding different equitizations may still cause mismatches. 

To fully resolve the allocation/deallocation-size mismatch, we might need to record the actual allocation size during each allocation, similar to how you store the allocation size for each allocated memory `p`. This way, regardless of what `m` was provided during `deallocate(p, m)`, we can always determine the actual allocation size from `p`. Your code provides a neat implementation of this approach.

> This is a lot more complicated but I think it works. If you agree with that direction, we could land this PR as-is and then tackle this additional change as a follow-up, since it can be done independently from this refactoring.

I agree it's getting a bit complicated but it seems worth proceeding. I agree with your direction. I'll keep the current refactoring in this PR as I really want the `increasing_allocator` utility in a single location, so that other work (including my own ongoing work) can use this class without duplicating the code. Follow-up work to further improve this class can be done independently after this PR.

[the remark this answers is not in the document]
Comment (winner245) [reply libcxx/test/support/increasing_allocator.h]: Maybe another way to avoid the size mismatch problem is to use operator new and delete, which eliminates the need to manually specify the deallocation size:

```cpp
template <typename T>
struct increasing_allocator {
  using value_type         = T;
  std::size_t min_elements = 1000;
  increasing_allocator()   = default;
  
#if TEST_STD_VER >= 23
  std::allocation_result<T*> allocate_at_least(std::size_t n) {
    if (n < min_elements)
      n = min_elements;
    min_elements += 1000;
    return {static_cast<T*>(::operator new(n * sizeof(T))), n};
  }
#endif // TEST_STD_VER >= 23

  T* allocate(std::size_t n) {
    return static_cast<T*>(::operator new(n * sizeof(T)));
  }

  void deallocate(T* p, std::size_t) TEST_NOEXCEPT { ::operator delete(p); }
};
```

**Review by ldionne — APPROVED** [review]
LGTM. Let's tackle the allocation/deallocation size mismatch separately.
